onnxruntime/core/providers/dml/OperatorAuthorHelper/OperatorHelper.cpp

```diff
@@ -550,6 +550,90 @@
+        std::vector<DimensionType>& outputShape)
+    {
+        // Get the padded input shapes, and undo the effect of padding removal from the output shape.
+        if (inputShape1.size() == 1)
+        {
+            inputShape1.push_back(1);
+            inputStride1.push_back(0);
+            outputShape.push_back(1);
+        }
+
+        if (inputShape0.size() == 1)
+        {
+            inputShape0.insert(inputShape0.begin(), 1);
+            inputStride0.insert(inputStride0.begin(), 0);
+            outputShape.insert(outputShape.end() - 1, 1);
+        }
+
+        auto broadcastedRank = std::max(inputShape0.size(), inputShape1.size());
+        inputShape0.insert(inputShape0.begin(), (broadcastedRank - inputShape0.size()), 1);
+        inputStride0.insert(inputStride0.begin(), (broadcastedRank - inputStride0.size()), 0);
+
+        inputShape1.insert(inputShape1.begin(), (broadcastedRank - inputShape1.size()), 1);
+        inputStride1.insert(inputStride1.begin(), (broadcastedRank - inputStride1.size()), 0);
+
+        // Broadcast the extra dimensions of each input, then add the truncated matrix dimensions.
+        BroadcastTensorShapeAndSetStrides(
+            gsl::make_span(inputShape0).subspan(0, broadcastedRank - 2),
+            gsl::make_span(inputStride0).subspan(0, broadcastedRank - 2),
+            gsl::make_span(inputShape1).subspan(0, broadcastedRank - 2),
+            gsl::make_span(inputStride1).subspan(0, broadcastedRank - 2)
+        );
+    }
+
+    std::pair<std::vector<uint32_t>, std::vector<uint32_t>> FusedMatMulSizeAndStride(
+        gsl::span<const uint32_t> sizes,
+        int32_t transBatch,
+        int32_t trans)
+    {
+        const uint32_t dimensionCount = sizes.size();
+        std::vector<uint32_t> newStrides(dimensionCount);
+        std::vector<uint32_t> newSizes(sizes.begin(), sizes.end());
+
+        // Calculate packed strides.
+        uint32_t stride = 1;
+        for (int i = dimensionCount - 1; i >= 0; i--)
+        {
+            newStrides[i] = stride;
+            stride *= sizes[i];
+        }
+
+        // According to contrib ops shape inference
+        // https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/graph/contrib_ops/contrib_defs.cc#L215
+        // `transBatch` needs to be applied first and then `trans`.
+        if (transBatch)
+        {
+            ML_CHECK_VALID_ARGUMENT(dimensionCount > 2, 
+                "FusedMatMul operator: Tensor size should be more than 2, if attribute transBatch is true");
+            uint32_t secondLastStride = newStrides[dimensionCount - 2];
+            uint32_t secondLastSize = newSizes[dimensionCount - 2];
+
+            for (int i = dimensionCount - 2; i > 0; i--)
+            {
+                newStrides[i] = newStrides[i - 1];
+                newSizes[i] = newSizes[i - 1];
+            }
+
+            newStrides[0] = secondLastStride;
+            newSizes[0] = secondLastSize;
+        }
```

Comment:
Is this whole thing really just an std::rotate?

[https://en.cppreference.com/w/cpp/algorithm/rotate](https://en.cppreference.com/w/cpp/algorithm/rotate)

`std::rotate(newStrides.begin(), newStrides.begin() + 1, newStrides.end());`

(I might have the +1 wrong)

Reply:
Learnt lot of c++ new functions today.